### PR TITLE
[process orchestrator] Use systemd if available

### PR DESCRIPTION
**Pull request by antiguru**
This PR uses systemd-run if available to limit the CPU and memory resources a process can consume. It detects at runtime if systemd is available, and defaults to the current behavior if it's not.

Also adjusts the replica size definitions to include a CPU limit, and introduce some variants with explicit memory limits. We could align this to be closer to what we use in production.

  * This PR adds a feature that has not yet been specified.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: src/adapter/src/catalog/config.rs
Comment (jkosh44) [inline src/adapter/src/catalog/config.rs:1]: Every time I see this function I'm a bit scared that the `default` method should never be called in production. I supposed that's not exactly relevant to this PR though.

**Review by jkosh44 — APPROVED** [review]
I know nothing about systemd, but the code LGTM

**Comment by nrainer-materialize** [comment]
I triggered a [nightly build](https://buildkite.com/materialize/nightlies/builds/4913#). Results look good (the platform checks are fixed on main with https://github.com/MaterializeInc/materialize/pull/22719; the postgres-inconsistency is unrelated, I will take care of that).

**Review by vmarcos — CHANGES_REQUESTED** [review]
This looks fine to me overall, but perhaps we'd like to abstract the formula to compute `workers -> CpuLimit` only once instead of writing it a few times. If I did not misread something, there might be one constant that needs fixing below?

File: src/adapter/src/catalog/config.rs [hunk 1]
```diff
@@ -171,7 +183,7 @@ impl Default for ClusterReplicaSizeMap {
                 format!("mem-{scale}"),
                 ReplicaAllocation {
                     memory_limit: Some(MemoryLimit(ByteSize(u64::cast_from(scale) * (1 << 30)))),
-                    cpu_limit: None,
+                    cpu_limit: Some(CpuLimit::from_millicpus(2000)),
```
[the remark this answers is not in the document]
Comment (antiguru) [reply src/adapter/src/catalog/config.rs]: Yup, makes sense! Added and fixed.

File: src/adapter/src/catalog/config.rs [hunk 2]
```diff
@@ -122,20 +122,30 @@ impl Default for ClusterReplicaSizeMap {
...
-                        disk_limit: None,
-                        scale: 1,
-                        workers: workers.into(),
-                        credits_per_hour: 1.into(),
-                        disabled: false,
-                    },
-                )
+                [
+                    (workers.to_string(), None),
+                    (format!("{workers}-8G"), Some(8)),
```
[the remark this answers is not in the document]
Comment (antiguru) [reply src/adapter/src/catalog/config.rs:138]: Good idea; done.

File: src/adapter/src/catalog/config.rs
Comment (antiguru) [reply src/adapter/src/catalog/config.rs:1]: Filed #22747 so we're less likely to forget. I've been thinking the same!

**Review by vmarcos — APPROVED** [review]
Nice, thank you!

**Pull request merged**